l2geth/rollup/sync_service.go

```diff
@@ -632,101 +486,171 @@ func (s *SyncService) SetLatestVerifiedIndex(index *uint64) {
+	}
+	return fmt.Errorf("Received tx at index %d when looking for %d", *index, next)
+}
 
-	log.Debug("Maybe applying transaction", "hash", tx.Hash().Hex())
+// applyHistoricalTransaction will compare a historical transaction against what
+// is locally indexed. This will trigger a reorg in the future
+func (s *SyncService) applyHistoricalTransaction(tx *types.Transaction) error {
+	if tx == nil {
+		return errors.New("Transaction is nil in applyHistoricalTransaction")
+	}
 	index := tx.GetMeta().Index
 	if index == nil {
-		return fmt.Errorf("nil index in maybeApplyTransaction")
+		return errors.New("No index is found in applyHistoricalTransaction")
 	}
-	// Handle off by one
+	// Handle the off by one
 	block := s.bc.GetBlockByNumber(*index + 1)
-
-	// The transaction has yet to be played, so it is safe to apply
 	if block == nil {
-		err := s.applyTransaction(tx)
-		if err != nil {
-			return fmt.Errorf("Maybe apply transaction failed on index %d: %w", *index, err)
-		}
-		return nil
+		return fmt.Errorf("Block %d is not found", *index+1)
 	}
-	// There is already a transaction at that index, so check
-	// for its equality.
 	txs := block.Transactions()
 	if len(txs) != 1 {
-		log.Info("block", "txs", len(txs), "number", block.Number().Uint64())
-		return fmt.Errorf("More than 1 transaction in block")
+		return fmt.Errorf("More than one transaction found in block %d", *index+1)
 	}
-	if isCtcTxEqual(tx, txs[0]) {
-		log.Info("Matching transaction found", "index", *index)
+	if !isCtcTxEqual(tx, txs[0]) {
+		log.Error("Mismatched transaction", "index", index)
 	} else {
-		log.Warn("Non matching transaction found", "index", *index)
+		log.Debug("Historical transaction matches", "index", *index, "hash", tx.Hash().Hex())
 	}
 	return nil
 }
 
-// Lower level API used to apply a transaction, must only be used with
-// transactions that came from L1.
-func (s *SyncService) applyTransaction(tx *types.Transaction) error {
+// applyTransactionToTip will do sanity checks on the transaction before
+// applying it to the tip. It blocks until the transaction has been included in
+// the chain.
+func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
+	if tx == nil {
+		return errors.New("nil transaction passed to applyTransactionToTip")
+	}
+	if tx.L1Timestamp() == 0 {
+		ts := s.GetLatestL1Timestamp()
+		bn := s.GetLatestL1BlockNumber()
+		tx.SetL1Timestamp(ts)
+		tx.SetL1BlockNumber(bn)
+	} else if tx.L1Timestamp() > s.GetLatestL1Timestamp() {
+		ts := tx.L1Timestamp()
+		bn := tx.L1BlockNumber()
+		s.SetLatestL1Timestamp(ts)
+		s.SetLatestL1BlockNumber(bn.Uint64())
+	} else if tx.L1Timestamp() < s.GetLatestL1Timestamp() {
+		// TODO: this should force a reorg
+		log.Error("Timestamp monotonicity violation", "hash", tx.Hash().Hex())
+	}
+
+	if tx.GetMeta().Index == nil {
+		index := s.GetLatestIndex()
+		if index == nil {
+			tx.SetIndex(0)
+		} else {
+			tx.SetIndex(*index + 1)
+		}
+	}
+	s.SetLatestIndex(tx.GetMeta().Index)
```

Comment:
Can it ever be the case that the received tx's index is smaller than the latest?

Reply:
It should never be the case due to calls higher up the stack, `applyIndexedTransaction` does an index comparison. The exception would be a reorg and then that would be desired behavior